### PR TITLE
inductor: rewrite mkldnn fx fusion using pattern_matcher(conv_unary)

### PR DESCRIPTION
**Pull request by XiaobingSuper**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #97141
* #97140
* #97139
* __->__ #97007

cc @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/pattern_matcher.py [hunk 1]
```diff
@@ -598,6 +598,224 @@ def addmm(match, mat1, mat2, inp):
...
 
 
+if torch._C.has_mkldnn:
+    mkldnn = torch.ops.mkldnn
+    # _conv_args = (Arg(), Arg(), Arg(), Arg(), Arg(), Arg(), Arg(), KeywordArg('attr'), Arg(),  Arg(),)
+    _conv_args = (Arg(), Arg(), Arg(), Arg(), Arg(), Arg(), Arg())
+    _user_1 = [CallFunction(aten.mkldnn_convolution, *_conv_args, _users=1)]
+    _user_2 = [CallFunction(aten.mkldnn_convolution, *_conv_args, _users=2)]
+    _user_3 = [CallFunction(aten.mkldnn_convolution, *_conv_args, _users=3)]
+    _user_4 = [CallFunction(aten.mkldnn_convolution, *_conv_args, _users=4)]
```
Comment (jgong5) [inline torch/_inductor/pattern_matcher.py:608]: Is `_users` only mandatory for `CallFunction` for computation_call or it is also needed for `CallFunction` of post-ops?

**Review by jgong5 — APPROVED** [review]
LGTM.

File: torch/_inductor/pattern_matcher.py (hunk 1 above)
Comment (XiaobingSuper) [reply torch/_inductor/pattern_matcher.py:608]: Only mandatory for CallFunction for computation_call.

File: torch/_inductor/mkldnn.py [hunk 2]
```diff
@@ -138,7 +130,7 @@ def _update_module_params(self, conv, unary, input_size):
 
     def _conv_forward(self, input, weight, bias):
         if self.padding_mode != "zeros":
-            return torch.ops.mkldnn._convolution_pointwise(
+            return torch.mkldnn_convolution(
```
Comment (EikanWang) [inline torch/_inductor/mkldnn.py:133]: Let's add a single `conv` and linear for the repacked `conv`. The semantic of `_conv_pointwise` is confusing as it serves both the single conv and the fused conv.

File: torch/_inductor/pattern_matcher.py (hunk 1 above)
Comment (EikanWang) [inline torch/_inductor/pattern_matcher.py:608]: I'm thinking of adding a utility function to deduce the value of `_users` directly, as Jiong suggested. The `_user_x` is confusing.

File: torch/_inductor/pattern_matcher.py [hunk 3]
```diff
@@ -598,6 +598,224 @@ def addmm(match, mat1, mat2, inp):
...
+            computation_args = [arg for arg in args]
+            if matched:
+                computation_args += ["hardtanh", [min_value, max_value], ""]
+                return L[computation_op](*computation_args)
+            else:
+                computation_args += ["none", [], ""]
+                conv_out = L[computation_op](*computation_args)
+                return L[aten.clamp_max](
+                    L[aten.clamp_min](conv_out, min_value), max_value
+                )
```
Comment (EikanWang) [inline torch/_inductor/pattern_matcher.py:805]: This is the decompose version of `hardtanh`, right? Is there any way to call the decomposed function directly but not duplicate it again.

File: torch/_inductor/pattern_matcher.py (hunk 1 above)
Comment (XiaobingSuper) [reply torch/_inductor/pattern_matcher.py]: changed.

File: torch/_inductor/pattern_matcher.py (hunk 3 above)
Comment (XiaobingSuper) [reply torch/_inductor/pattern_matcher.py:851]: I didn't find a way to do it.

File: torch/_inductor/mkldnn.py [hunk 4]
```diff
@@ -93,11 +93,10 @@ def check_binary_op_kwargs_is_default(node):
     return True
 
 
-class ConvUnary2d(nn.Conv2d):
+class PackedConv2d(nn.Conv2d):
```
Comment (jgong5) [inline torch/_inductor/mkldnn.py:96]: A side note: we leave weight repacking before AOT autograd while move fusion after it. In the future, we may consider to move weight prepacking after AOT autograd too and do it together with constant folding passes there.

**Review by jgong5 — APPROVED** [review]
LGTM

**Review by EikanWang — APPROVED** [review]

File: torch/_inductor/pattern_matcher.py [hunk 5]
```diff
@@ -608,6 +608,279 @@ def addmm(match, mat1, mat2, inp):
         return L[aten.add](inp, L[aten.mm](mat1, mat2))
 
 
+# TODO(XiaobingSuper): move it to fx_passes/mkldnn_fuse.py
```
Comment (desertfire) [inline torch/_inductor/pattern_matcher.py:611]: I am ok with this stack of changes, but please move them into a separate file.

Comment (jansel) [reply torch/_inductor/pattern_matcher.py:611]: You can create a `fx_passes/` folder to match with https://github.com/pytorch/pytorch/pull/97741

File: torch/_inductor/pattern_matcher.py [hunk 6]
```diff
@@ -608,6 +608,279 @@ def addmm(match, mat1, mat2, inp):
...
+        )
+
+    def relu_fusion(computation_call):
+        return CallFunction(aten.relu, computation_call)
+
+    def sigmoid_fusion(computation_call):
+        return CallFunction(aten.sigmoid, computation_call)
+
+    def tanh_fusion(computation_call):
+        return CallFunction(aten.tanh, computation_call)
```
Comment (jansel) [inline torch/_inductor/pattern_matcher.py:729]: You can match multiple functions in a single pattern like:
```suggestion
    def combined_fusion(computation_call):
        return CallFunction([aten.relu, aten.sigmoid, aten.tanh], computation_call)
```

Then figure out which one you hit with the `match` arg to the lowering.

This should allow you to have fewer patterns throughout the stack.

File: torch/_inductor/pattern_matcher.py (hunk 5 above)
Comment (XiaobingSuper) [reply torch/_inductor/pattern_matcher.py]: Yes, move it to a separate file.

File: torch/_inductor/pattern_matcher.py (hunk 6 above)
Comment (XiaobingSuper) [reply torch/_inductor/pattern_matcher.py]: yes, it is simplified.

**Comment by XiaobingSuper** [comment]
@jansel @desertfire, please help review this code again. Thanks!

File: torch/_inductor/fx_passes/mkldnn_fusion.py [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+        _leaky_relu_patterns = [
+            _leaky_relu_fusion(user) for user in _computation_user_3
+        ]
+        _hardtanh_patterns = [_hardtanh_fusion(user) for user in _computation_user_1]
+        for pattern, computation_op in zip(_leaky_relu_patterns, computation_ops):
+            _register_leaky_relu_fusion_lowering(pattern, computation_op)
+        for pattern, computation_op in zip(_hardtanh_patterns, computation_ops):
+            _register_hardtanh_fusion_lowering(pattern, computation_op)
+
+    def _mkldnn_fusion_init():
```
Comment (jansel) [inline torch/_inductor/fx_passes/mkldnn_fusion.py:258]:
```suggestion
    @functools.lru_cache(None)
    def _mkldnn_fusion_init():
```
So this code only runs once.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/fx_passes/mkldnn_fusion.py (hunk 7 above)
Comment (XiaobingSuper) [reply torch/_inductor/fx_passes/mkldnn_fusion.py:261]: changed.

File: torch/_inductor/fx_passes/post_grad.py [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+
+    if config.reordering:
+        # has some issues with mutation in inference mode
+        reorder_for_locality(gm.graph)
+
+    if config.pattern_matcher:
+        if torch._C.has_mkldnn:
+            from .mkldnn_fusion import _mkldnn_fusion_init
+
+            _mkldnn_fusion_init()
```
[the remark this answers is not in the document]
Comment (XiaobingSuper) [reply torch/_inductor/fx_passes/post_grad.py]: Moved the code to lazy_init(). Thanks!

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**